spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSource.scala

```diff
@@ -0,0 +1,516 @@
+ * actions in a Delta log.
+ * For proper offset tracking(move the offset to the next version if all data is consumed in the
+ * current version), there are also special sentinel values with index = -1 and add = null.
+ *
+ * This class is not designed to be persisted in offset logs or such.
+ *
+ * @param version The version of the Delta log containing this AddFile.
+ * @param index The index of this AddFile in the Delta log.
+ * @param add The AddFile.
+ * @param isLast A flag to indicate whether this is the last AddFile in the version. This is used
+ *               to resolve an off-by-one issue in the streaming offset interface; once we've read
+ *               to the end of a log version file, we check this flag to advance immediately to the
+ *               next one in the persisted offset. Without this special case we would re-read the
+ *               already completed log file.
+ */
+private[sharing] case class IndexedFile(
+  version: Long,
+  index: Long,
+  add: AddFile,
+  isLast: Boolean = false) {
+
+  def getFileAction: AddFile = { add }
+}
+
+/**
+ * Base trait for the Delta Sharing Source, that contains methods that deal with
+ * getting changes from the delta sharing server.
+ * TODO(lin.zhou) Support SupportsTriggerAvailableNow
+ */
+trait DeltaSharingSourceBase extends Source
+  with SupportsAdmissionControl
+  with Logging { self: DeltaSharingSource =>
+
+  val snapshot: RemoteSnapshot = deltaLog.snapshot()
+
+  override val schema: StructType = {
+    val schemaWithoutCDC = snapshot.schema
+    if (options.readChangeFeed) {
+      throw DeltaSharingErrors.CDFNotSupportedInStreaming
+    } else {
+      schemaWithoutCDC
+    }
+  }
+
+  protected var sortedFetchedFiles: Seq[IndexedFile] = Seq.empty
+
+  protected var lastGetVersionTimestamp: Long = -1
+  protected var lastQueriedTableVersion: Long = -1
+  private val QUERY_TABLE_VERSION_INTERVAL_MILLIS = 30000 // 30 seconds
+
+  protected def intervalGetTableVersion: Long = {
+    val currentTimeMillis = System.currentTimeMillis()
+    if (lastGetVersionTimestamp == -1 ||
+      (currentTimeMillis - lastGetVersionTimestamp) >= QUERY_TABLE_VERSION_INTERVAL_MILLIS) {
+      lastQueriedTableVersion = deltaLog.client.getTableVersion(deltaLog.table)
+      lastGetVersionTimestamp = currentTimeMillis
+    }
+    lastQueriedTableVersion
+  }
+
+  /**
+   * Fetch the changes from delta sharing server starting from (fromVersion, fromIndex).
+   * The start point should not be included in the result, it should be consumed in the last batch.
+   *
+   * If sortedFetchedFiles is not empty, return directly.
+   * Else, fetch file changes from the delta sharing server and store them in sortedFetchedFiles.
+   */
+  protected def getFileChanges(
+    fromVersion: Long,
+    fromIndex: Long,
+    isStartingVersion: Boolean): Unit = {
+    if (!sortedFetchedFiles.isEmpty) { return }
+
+    if (fromVersion > intervalGetTableVersion) {
+      return
+    }
```

Comment:
Question: The getFiles call below may return files from a version which may be greater than intervalGetTableVersion... right? Is that ok? Consider adding some comments.